pkg/filetree/filetree.go

```diff
@@ -0,0 +1,754 @@
+	//      we return NIL.
+	//  - /other/place/here DOES NOT need to exist in the Tree, it either
+	//          a) exists as a regular file --in which case return the discovered file.Reference
+	//	        b) does not exist --return NIL
+	//          c) or exists as a symlink that may or may not resolve --this last case does not matter since the
+	//             PATH has been resolved to a file.Reference, which is all that matters)
+	//
+	// Therefore we can safely lookup the path first without worrying about symlink resolution yet... if there is a
+	// hit, return it! If not, fallback to symlink resolution.
+
+	currentNode, err := t.node(path, linkResolutionStrategy{})
+	if err != nil {
+		return false, nil, err
+	}
+	if currentNode != nil && (!currentNode.IsLink() || currentNode.IsLink() && !userStrategy.FollowBasenameLinks) {
+		return true, currentNode.Reference, nil
+	}
+
+	// symlink resolution!... within the context of container images (which is outside of the responsibility of this object)
+	// the only really valid resolution of symlinks is in squash trees (both for an image and a layer --NOT for trees
+	// that represent a single union FS layer.
+	currentNode, err = t.node(path, linkResolutionStrategy{
+		FollowAncestorLinks:          true,
+		FollowBasenameLinks:          userStrategy.FollowBasenameLinks,
+		DoNotFollowDeadBasenameLinks: userStrategy.DoNotFollowDeadBasenameLinks,
+	})
+	if currentNode != nil {
+		return true, currentNode.Reference, err
+	}
+	return false, nil, err
+}
+
+func (t *FileTree) node(p file.Path, strategy linkResolutionStrategy) (*filenode.FileNode, error) {
+	if !strategy.FollowLinks() {
+		n := t.tree.Node(filenode.IDByPath(p))
+		if n == nil {
+			return nil, nil
+		}
+		return n.(*filenode.FileNode), nil
+	}
+
+	var currentNode *filenode.FileNode
+	var err error
+	if strategy.FollowAncestorLinks {
+		currentNode, err = t.resolveAncestorLinks(p)
+		if err != nil {
+			return currentNode, err
+		}
+	} else {
+		n := t.tree.Node(filenode.IDByPath(p))
+		if n != nil {
+			currentNode = n.(*filenode.FileNode)
+		}
+	}
+
+	// link resolution has come up with nothing, return what we have so far
+	if currentNode == nil {
+		return currentNode, nil
+	}
+
+	if strategy.FollowBasenameLinks {
+		currentNode, err = t.resolveNodeLinks(currentNode, !strategy.DoNotFollowDeadBasenameLinks)
+	}
+	return currentNode, err
+}
+
+// return FileNode of the basename in the given path (no resolution is done at or past the basename).
+func (t *FileTree) resolveAncestorLinks(path file.Path) (*filenode.FileNode, error) {
+	var pathParts = strings.Split(string(path.Normalize()), file.DirSeparator)
+	var currentPathStr string
+	var currentPath file.Path
+	var currentNode *filenode.FileNode
+	var err error
+
+	// iterate through all parts of the path, replacing path elements with link resolutions where possible.
+	for idx, part := range pathParts {
+		if (part == "" || part == file.DirSeparator) && idx == 0 {
```

Comment:
It seems that these conditional checks would be equal to checking if the `part == ""`. The rest of the conditions read somewhat redundant. Are they really necessary? Or in what situations the root path `"/"` would produce something other than an empty string?